### ISSUE TITLE
RFC: Introduce a keyword for map/var definitions/declarations.

### ISSUE THREAD

**Issue by jordalgo** [bpftrace/bpftrace#3423]
This issue is just to consolidate some discussions here: https://github.com/bpftrace/bpftrace/pull/3367, https://github.com/bpftrace/bpftrace/issues/2954

Doing a bit of research, here is how other languages break down in terms of keywords on the left side of variable declarations/initializations.

## Languages that don't have any space-separate label/keyword:
- Python
- Ruby (global variables are prefixed with `$`)
- R
- Haskell
- Perl (scalar variables are prefixed with `$`)
- PHP/Hack (prefixed with `$`)
- C / C++ / C# / Java (require a type on the left)

## Languages that have a label/keyword
- Clojure (`def`)
- Go (`var`)
- Lua  / Bash (global variables have no label but local variables have `local` label)
- Scala / Kotlin (`var` and `val`)
- Dart (has `var` but also can skip that by specifying the type before the variable name)
- Zig (`var` and `const`)
- Rust (`let` and `const`)
- Javascript ((`let` and `const` - `var` is deprecated)
- Swift (`let` and `var`, though the latter is mutable)

**For a quick vote (but feel free to comment):**
Thumbs up this comment if you are in favor of ADDING a new label/keyword 👍 
Thumbs down this comment if you are in favor of NOT ADDING a new label/keyword 👎

**Comment by fbs** [comment]
Thanks for opening this. Useful to get a consistent experience around all types of variables.

As we already have `$` and `@` to identify variables I don't think we have a need for let from a parsing perspective. One place where `let` can be useful is if we want to introduce [variable shadowing](https://en.wikipedia.org/wiki/Variable_shadowing) in scopes. E.g.

```
probe {
  $x = 1
  for (...) {
     $x = "abc"
     print($x)
  }
}
```

This would be a type error right now. Although we can probably also get away without shadowing?

**Comment by jordalgo** [comment]
> One place where let can be useful is if we want to introduce [variable shadowing](https://en.wikipedia.org/wiki/Variable_shadowing) in scopes

I think even then I might prefer the keyword `local` as `let` seems to be a bit ambiguous across languages.

**Comment by viktormalik** [comment]
Agreed with the argument of @fbs here.

**Comment by jordalgo** [comment]
> It doesn't have to be let, var or other options are ok too. My point was more that the let allows you to differentiate between an assignment (which might implicitly define a var) and a new declaration, i.e. 'in this scope i want a var of type x, optionally with a value of y'. Maybe we should rephrase the RFC into `introduce a keyword for map/var definitions?'

Good point. I'll update the RFC and make it a bit more generic.

**Comment by jordalgo** [comment]
> Although we can probably also get away without shadowing?

I find variable shadowing confusing and don't think we need it or should allow it.

**Comment by JakeHillion** [comment]
I think there are three things that we're quite explicitly trading off with the syntax here:
1. Terseness. How few characters can we use?
2. New syntax. Do I have to remember a new addition to the language or does it feel familiar? (particularly this comment: https://github.com/bpftrace/bpftrace/issues/3360#issuecomment-2302425624)
3. Readability. With no context of the bpftrace language, can you understand what the script does?

To me, terseness is unimportant here. Terse scripts are enabled by competent type inference/keyword-free auto (in C++ terminology) that we already have. By the time we require type annotations we should be firmly focused on the other two: can we reuse existing syntax in a way users will understand, and how readable is the final result.

My opinion: more keywords create more readable code.

I like `let`, even if some of the usage might feel a bit unusual. It's short and makes for easy switching of the parsing "context" (not sure that's the right word) and then we can allow a lot of syntax variations afterwards. I see it expanding to things like:
```
let global @x = ArrayMap;               // map implementation specified, type inferred as usual
let global @x: uint32 = ArrayMap;  // map type and implementation specified
let global @x: [uint32 -> uint64];   // map type (a mapping map) specified
let $daddr : inet_t;                           // local variable type specified (implied by the dollar)

let type MyTypedef = uint64;  // a typedef?
let @x: uint32;                           // global inferred from the `@`?
let global x: uint32;                   // `@` (mappiness) inferred from the `global`?

let global @x = 0;                     // initial value for a map?
let global @x = ArrayMap{1};  // initial value for a map with specified implementation?
let global @x = {0->1, 2->3};  // a pair of initial mappings specified?
```

I think the excess of keywords in most of these statements makes them more readable to someone who hasn't seen the syntax before than alternatives. I also think they achieve a nice balance with terseness.

That's my 2 cents. Other short tokens are available in place of let :)

**Comment by ajor** [comment]
> ## Languages that don't have any space-separate label/keyword:
> * Python
> * Ruby (global variables are prefixed with `$`)
> * R
> * Haskell
> * Perl (scalar variables are prefixed with `$`)
> * PHP/Hack (prefixed with `$`)
> * C / C++ / C# / Java (require a type on the left)

As far as I can tell, Ruby, R, Perl and PHP/Hack don't support variable declarations. They're similar to our current behaviour, where variables must be initialised on first use. Haskell's an odd one with very different variables (and no declarations from what I understand). Python did introduce declarations in a relatively recent version, but they are ignored by the interpreter and rarely used.

This suggests that it's far more common for languages to use some sort of keyword for variable declarations, for what it's worth.

**Comment by viktormalik** [comment]
Reading the discussion here, I'd like to clarify some things.

Are we considering adding a *mandatory* keyword for all map/var definitions/declarations? That is, something that would break essentially every script out there? To me, this seems to be a major compatibility breakage and I don't really see an equivalently large benefit that we would gain in exchange.

> To me, terseness is unimportant here. Terse scripts are enabled by competent type inference/keyword-free auto (in C++ terminology) that we already have. By the time we require type annotations we should be firmly focused on the other two: can we reuse existing syntax in a way users will understand, and how readable is the final result.

I beg to disagree here. I'd say that one-liners are still a frequent (perhaps even the most frequent?) way of using bpftrace and to them, terseness matters. Adding several characters to each variable def will make the one-liner much longer and inherently less readable.

Also, I was under impression that we will only require type annotations for declarations. We could certainly add the keyword there but I don't see a reason of doing that since the parser doesn't need it.

To sum it up, I don't think that adding a new mandatory keyword would be a good idea here. I'd probably be fine with adding an optional keyword (or make it mandatory just for declarations) but there seems to be no value in that.

**Comment by jordalgo** [comment]
I agree with @viktormalik here. I think if we add `let` it needs to have a purpose/function. In all the examples above
it doesn't seem to, which not only makes the code less terse but less readable.

In the popular languages listed `let`/`var`/`val` are defining a property of the variable e.g. they are mutable or not, global or local, etc. bpftrace only has two kinds of variables (global and scratch) and they are already distinguishable from the `$` or `@` prefixes. I think if folks want to make a case for replacing these prefixes with keywords to make them more understandable to new users, that's a separate discussion (though this would hurt one-liners, which is not good).

Additionally, adding `let` or `var` or another keyword doesn't let you omit other portions of a variable or map definition/declaration e.g. `$a: uint16;` vs `let $a: uint16;` or `@x: [uint32 -> uint64];` vs `let global @x: [uint32 -> uint64];`. I don't know what `global` or `let` are adding in terms of information conveyed to the user or to bpftrace.

**Comment by jordalgo** [comment]
> This suggests that it's far more common for languages to use some sort of keyword for variable declarations, for what it's worth.

You mean except for C, C++, Java 😄  The only difference is that they put the type on the left and bpftrace will have it on the right. Also, I think there needs to be an underlying reason for adding something new to the language other than it's common or is in this other language.

**Comment by ajor** [comment]
> Are we considering adding a _mandatory_ keyword for all map/var definitions/declarations?

No definitely not.

Variable declarations are purely an extension and we won't be getting rid of the `$x = 123` syntax, which creates a new variable while assigning to it. As you say, terseness does matter for one-liners. But I imagine that explicitly declared variables won't generally be used in one-liners - they'll be mostly reserved for larger scripts where terseness isn't important, but readability is.

> Also, I think there needs to be an underlying reason for adding something new to the language other than it's common or is in this other language.

Other languages using a particular syntax isn't a reason by itself to copy the syntax ourselves, but it does carry weight to me for a couple of reasons:
- Familiarity. BpfScript isn't going to be anyone's primary programming language - it'll be something extra to learn to solve a specific problem. For that reason, making the syntax similar to existing languages, which users will already be familiar with, will make it easier for them to pick up and write.
- They've generally put more thought into these things than we have! There might be logical reasons behind their syntax choices that we haven't discovered yet.

> I think if folks want to make a case for replacing these prefixes with keywords to make them more understandable to new users, that's a separate discussion

Another related discussion is how we're going to expose variables and maps imported from externally created modules/libraries: #2976. i.e. if a BPF-C program creates a global variable called `blah`, do we expose it in BpfScript without the `$`/`@` prefix?

I'd say that it's neither of these is something that we need to discuss in detail upfront though. I'm in favour of including a keyword because of the extensibility / future-proofing it provides. It'd allow for future extensions like these, those @JakeHillion described, or others we haven't yet though of, without introducing parsing ambiguities.

**Comment by viktormalik** [comment]
> > Are we considering adding a _mandatory_ keyword for all map/var definitions/declarations?
> 
> No definitely not.
> 
> Variable declarations are purely an extension and we won't be getting rid of the `$x = 123` syntax, which creates a new variable while assigning to it. As you say, terseness does matter for one-liners. But I imagine that explicitly declared variables won't generally be used in one-liners - they'll be mostly reserved for larger scripts where terseness isn't important, but readability is.

Ok, in that case, I've no strong opinion on using the `let` keyword for variable declarations. At the moment it doesn't seem necessary but as you point out, we may want to add language features where the new keyword would make the parsing much simpler.

So, no problem from my side with including it. I'd still disallow variable shadowing though.

**Comment by jordalgo** [comment]
> I'm in favour of including a keyword because of the extensibility / future-proofing it provides. It'd allow for future extensions like these, those @JakeHillion described, or others we haven't yet though of, without introducing parsing ambiguities.

Do we have some concrete examples of how it adds future-proofing / extensibility? Maybe this is the part I'm missing.

> Familiarity. BpfScript isn't going to be anyone's primary programming language - it'll be something extra to learn to help with a specific problem. For that reason, making the syntax similar to existing languages, which users will already be familiar with, will make it easier for them to pick up and write.

I think the examples at the beginning of this RFC show that there really isn't a convergence on the "right" way to do this from language to language. Also I have a feeling (though I don't have any data) that the majority of bpftrace users are systems engineers who have strong familiarity with C/C++ and if we cared about making the language more familiar to them then we'd consider more strongly not having `let` or `var` and placing the type on the left. User defined functions haven't landed yet (https://github.com/bpftrace/bpftrace/pull/3068) - it's not to late to revisit the decision to have variable types on the left (just sayin').

All this being said, I'm not against `let` or `var` as a permanent decision, I just want us to add them only when there is no other option or if they serve some very compelling purpose (like perhaps with the libraries feature). I just haven't seen that yet. They surely aren't necessary or provide any value for variable or map declarations from what I can see.

**Comment by viktormalik** [comment]
> Do we have some concrete examples of how it adds future-proofing / extensibility? Maybe this is the part I'm missing.

I can think of a couple of cases:

### 1. Reusing the assignment operator in declarations

Here is one case inspired by @JakeHillion's examples:
```
let @x: uint64_t = ArrayMap;
```

Here, the `let` keyword clearly says that this is a declaration and therefore the part after `=` can be treated as the implementation type to use. Without `let`, the `=` is what differentiates declarations from definitions and we'd have to come with a different syntax, e.g. using attributes:
```
[maptype=ArrayMap]
@x: uint64_t;
```

I think that the differentiation by the `let` keyword is more readable than by `=` although the latter would also allow to specify the implementation type on definitions which the former doesn't.

### 2. Declarations with auto-inferred type

If we ever wanted to allow these (I personally don't like them much), they look better with the `let` keyword:
```
let @x;
@x;
```
Of course, we could introduce the `auto` keyword and then the latter is fine, too:
```
let @x: auto;
@x: auto;
```
so this one is probably not much of an issue.

### 3. Global variables

We haven't thought about allowing global variables, yet. #3304 added support for them but for now, we only use them internally. The keyword would allow us to define global variables without an additional prefix:
```
let var: uint64_t;
```
although having non-prefixed variables would probably still complicate the parser quite a lot.

### Summary

Overall, I don't think that either of these is significant enough so to me, the only thing that remains is the "readability" where the new keyword makes spotting declarations in the code easier. I like it but I'm not sure if that's enough of a reason.

Also one final note - we'll have to decide if we want to use the keyword for function parameters, too, as they are declarations in fact. So, we would end up with something like:
```
fn add(let x: uint64_t, let y: uint64_t): uint64_t { ... }
```
which I don't like much.

**Comment by jordalgo** [comment]
Coming back to this conversation as I want to unblock lexical scoping, which is semi-blocked on variable declarations (cc @fbs @viktormalik). How do folks feel about moving forward with basic variable declarations without a keyword? I figure we could always add a new keyword later if that's what gets decided - though it seems like we're pretty split on the addition based on the conversation above.

The basic variable declaration would look like this:
```
$a: int64;
```
With initialization:
```
$a: int64 = 1;
```

Here is the draft implementation: https://github.com/jordalgo/bpftrace/commit/2bcb96a35287e39fb20e12cfb121a1e334474aab

**Comment by viktormalik** [comment]
After thinking this through, I don't have strong feelings about neither of these. Both variants (with and without the keyword) have their pros and cons and I feel like they are pretty balanced.

**Comment by ajor** [comment]
I'd still go with a keyword, for the extensibility reasons previously discussed, but I also think this point raised by @viktormalik is actually pretty important:

> ### 2. Declarations with auto-inferred type
> If we ever wanted to allow these (I personally don't like them much), they look better with the `let` keyword:
> 
> ```
> let @x;
> @x;
> ```

Going back to the reason we're thinking about this work, we've got this script:
https://github.com/bpftrace/bpftrace/blob/4bc94a15025f6c60b8cff20daea3d47e7891a18a/tools/tcpconnect.bt#L47-L53

How would we rewrite this with block scoping and variable declarations?
```
$daddr : WHAT_TYPE_GOES_HERE?
if ($inet_family == AF_INET) {
  $daddr = ntop($sk->__sk_common.skc_daddr);
} else {
  $daddr = ntop($sk->__sk_common.skc_v6_daddr.in6_u.u6_addr8);
}
```
(I think the answer is `inet_t[16]`, but it's not obvious)

So we'd probably need to include some form of type deduction for explicitly declared variables to make them usable with builtin types. i.e.:
```
let $daddr;
```

**Comment by ajor** [comment]
> I think the examples at the beginning of this RFC show that there really isn't a convergence on the "right" way to do this from language to language.

I missed this before, but I disagree. I think the examples in the first post are misleading. [When I went through the languages](https://github.com/bpftrace/bpftrace/issues/3423#issuecomment-2315771062), it turned out that there was only one that actually supports variable declarations without a keyword: Python, and even then they're ignored by the interpreter.

> Also I have a feeling (though I don't have any data) that the majority of bpftrace users are systems engineers who have strong familiarity with C/C++ and if we cared about making the language more familiar to them then we'd consider more strongly not having `let` or `var` and placing the type on the left

Fair point. I wasn't really involved in those discussions so can't comment.

**Comment by ajor** [comment]
What's the main benefit of having no keyword? Terseness, or something else I'm missing? If it's just for conciseness, then I think we need to think about the scenarios variable declarations are going to be used it. Do we see them being used in one-liners? From my perspective, they're something that's going to be used only in larger scripts where terseness doesn't matter.

> I think the examples at the beginning of this RFC show that there really isn't a convergence on the "right" way to do this from language to language.

I missed this before, but I disagree. I think the examples in the first post are misleading. [When I went through the languages](https://github.com/bpftrace/bpftrace/issues/3423#issuecomment-2315771062), it turned out that there was only one that actually supports variable declarations without a keyword: Python, and even then they're ignored by the interpreter.

> Also I have a feeling (though I don't have any data) that the majority of bpftrace users are systems engineers who have strong familiarity with C/C++ and if we cared about making the language more familiar to them then we'd consider more strongly not having `let` or `var` and placing the type on the left

Fair point. I wasn't really involved in those discussions so can't comment.

**Comment by jordalgo** [comment]
> So we'd probably need to include some form of type deduction for explicitly declared variables to make them usable with builtin types. i.e.:

Ok, this is a good reason for the additional keyword. I agree that just `$daddr;` looks wrong.

> it turned out that there was only one that actually supports variable declarations without a keyword

Why aren't we considering C/C++ part of this group that don't require a keyword? Because of `auto`? If so why not follow up on @viktormalik's suggestion for type inference by using `auto` e.g. `$daddr: auto;`. This seems like an additional keyword that feels more appropriate for bpftrace and can be re-used for user-defined function definitions.

> I wasn't really involved in those discussions so can't comment.

I'm still a bit confused here. Why is it set in stone that types need to go on the right instead of the left? Has something else been merged into trunk or release that means we can't revisit this? If types were on the left e.g. `uint8 $daddr;` or `auto $daddr;` - does that change things?

As an aside, I'd like to hear from the other maintainers on this topic 😄  - @danobi @fbs

**Comment by viktormalik** [comment]
> I'm still a bit confused here. Why is it set in stone that types need to go on the right instead of the left? Has something else been merged into trunk or release that means we can't revisit this?

I don't think it's set in stone but #2624 already implemented the `$var: type` variant for function parameters. It was discussed [here](https://github.com/bpftrace/bpftrace/pull/2624#discussion_r1264738941), the main argument was that the function return type is specified after colon (`fn name(...): type`) so the same should be used for parameters.

> If types were on the left e.g. `uint8 $daddr;` or `auto $daddr;` - does that change things?

To me, this doesn't make sense w.r.t. other parts of the language. In many ways, bpfscript is closer to scripting languages rather than languages derived from C (prefixing variables with `$`/`@`, default automatic type inference, ...) so I'm in favor of keeping the current `$var: type` syntax.

**Comment by JakeHillion** [comment]
Summarising a lot of in person discussion at LPC. Please let me know if I missed anything! I should have taken more notes...

### Declaring and initalising a variable with type inference
```
$x = 11;
```
This is good. We'd already established this but it won't be going anywhere. It declares the variable `$x` and initializes it to 11. Type inference is used. The argument for consistency made here requires a sentence along these lines in the documentation:

    `$x = VAL;` where `$x` is a previously unused scratch variable is defined as `let $x = VAL;`.

### Variable declaration with explicit type
```
$x: int8;
let $x: int8;
```
Next we move onto the new syntax required for block scoping: declaring a variable `x` with its type without initialising it, and including its type. We can do this with or without the `let`. We did not discuss types on the right vs types on the left, I believe this is settled as types on the right. As we require `$x` to follow the reasoning above, being an alias to a type of let when declaring a variable, we settle on the longer `let $x: int8;` syntax.

### Variable declarations with type inference
```
$x: auto;
let $x: auto;
let $x;
```
C/C++ use the auto keyword on the left, whereas we'd use it on the right as shown in the first line. Carbon gets the closest to line 2, allowing syntax like `var x: auto;`. The latter omits this `auto` for a terse syntax.

We avoided the `$x: auto;` for the reasons above - the succint `$x` syntax should be an easily described subset of the `let $x...;` syntax. We settled on the latter syntax (`$let x;`) for the consistency reasons as the `auto` on the right feels unnecessary.

### Variable declaration with explicit type and initialisation
```
$x: int8 = 4;
$x = (int8)4;
let $x: int8 = 4;
```
The first syntax was seen as unnecessary because the second cast syntax is already valid. The final syntax will be the full valid `let $x: int8 = 4;` syntax.

## Summary
### Existing syntax that remains valid
```
$x = 11;
$x = (uint8)19;
```

### New syntax
```
let $x;
let $x: uint8;
let $x = 11;
let $x: uint8 = 7;
```
That is,
```
let $x[: TYPE][ = VALUE];
```

We keep the existing syntaxes for variable initialization with implicit declaration, including the cast for explicit type. We add the new more verbose `let` syntax in order to expand this, including a version that covers the existing declaration/initialization version.

**Comment by jordalgo** [comment]
@JakeHillion Thanks for the write up! 

Adding one more note from the discussion (which may be in some of the other comments on this thread):
- let will be the added keyword but additional modifiers like "const" or "type", which will get fleshed out in a future RFC and used like `let const $x` or `let type blah = int64;`, will not be keywords

**Comment by jordalgo** [comment]
Just so it's codified in github comments, I'm good with adding `let` as described in @JakeHillion's comment above. 

As per our new language decision guidelines, can we get an approve/reject comment from the other maintainers: @ajor @fbs  @viktormalik @danobi

**Comment by danobi** [comment]
I think what @JakeHillion has laid out makes a lot of sense to me.

**Comment by viktormalik** [comment]
Approve from my side. I'm happy that we converged on this!

**Comment by jordalgo** [comment]
https://github.com/bpftrace/bpftrace/pull/3461

**Comment by jordalgo** [comment]
Closing this. We can discuss implementation details on the PR.
